tests/test_spidermiddleware_urllength.py

```diff
@@ -20,22 +20,36 @@ def setUp(self):
         self.stats = crawler.stats
         self.mw = UrlLengthMiddleware.from_settings(settings)
 
+        self.maxlength_infinite = 0
+        settings_infinite = Settings({'URLLENGTH_LIMIT': self.maxlength_infinite})
+
+        crawler_infinite = get_crawler(Spider)
+        self.spider_infinite = crawler_infinite._create_spider('foo_infinite')
+        self.stats_infinite = crawler_infinite.stats
+        self.mw_infinite = UrlLengthMiddleware.from_settings(settings_infinite)
         self.response = Response('http://scrapytest.org')
+
         self.short_url_req = Request('http://scrapytest.org/')
         self.long_url_req = Request('http://scrapytest.org/this_is_a_long_url')
         self.reqs = [self.short_url_req, self.long_url_req]
 
     def process_spider_output(self):
         return list(self.mw.process_spider_output(self.response, self.reqs, self.spider))
 
+    def process_spider_output_infinite(self):
+        return list(self.mw_infinite.process_spider_output(self.response, self.reqs, self.spider_infinite))
+
     def test_middleware_works(self):
         self.assertEqual(self.process_spider_output(), [self.short_url_req])
+        self.assertEqual(self.process_spider_output_infinite(), self.reqs)
 
     def test_logging(self):
         with LogCapture() as log:
             self.process_spider_output()
+            self.process_spider_output_infinite()
 
         ric = self.stats.get_value('urllength/request_ignored_count', spider=self.spider)
         self.assertEqual(ric, 1)
-
         self.assertIn(f'Ignoring link (url length > {self.maxlength})', str(log))
+        ric = self.stats_infinite.get_value('urllength/request_ignored_count', spider=self.spider_infinite)
+        self.assertIsNone(ric)
```

Comment:
I think this test method should remain unchanged, and a new one should be created.

Also, best tests usually test the limits where behavior change, and in this case I think it would be best to also test that, if the max length is set to 1, all requests are filtered out.

In a similar fashion, and this is _not_ a requirement since it’s not strictly related to this pull request, but I think the existing tests could be refactored to:
- Have a single URL instead of 2.
- Test the maximum max length that results in the request being filtered out (length of the request URL, minus 1).
- Test the minimum max length that results in the request not being filtered out (exact length of the request URL).

So, we would be testing the following limits: 0 (no filtering), 1 (filtering), URL length - 1 (filtering) URL length (no filtering).

Reply:
I think I can apply the refactor for the tests but I will need the previous questions to be answered because if I disable the middleware at settings level(Using [this](https://docs.scrapy.org/en/latest/topics/spider-middleware.html) documentation as reference), I will have to deal with a `None` middleware object in the tests so it will be impossible to use process_spider_output method for URLLENGTH_LIMIT = 0 case. Probably I am confused about the 'disable middleware' term meaning, so please clarify it to me

Reply:
With all these information, I think we can set three states here:
1) `URLLENGTH_LIMIT = 0`: For this one we could use the disable method from the settings. I think this is more like an special configuration than an empty one, so raising a NotConfigured exception would be confusing. In this case, we have to test that any time we create a UrlLengthMiddleware with from_settings it will return None.
2) `URLLENGTH_LIMIT != 0`: Normal behavior
3) `URLLENGTH_LIMIT` doesn't exist in settings: Maybe for this case we can raise a `NotConfigured` exception. This action could be done when some users do something that make impossible to load `URLLENGTH_LIMIT`(ex. del method or a buggy update of the settings). For this case, we have to test against the raising of a `NotConfigured` exception.

Does this make sense to you?